glean-core/ios/Glean/Net/PingUploader.swift

```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+typealias HeadersList = [String: String]
+
+/// The interface defining how to send pings.
+public protocol PingUploader {
+    /**
+     * Synchronously upload a ping to a server.
+     *
+     * @param request the ping upload request, locked within a uploader capability check
+     *
+     * @param callback used to return the status code of the upload response, so Glean knows whether or not to try again
+     */
+    func upload(
+        request: CapablePingUploadRequest,
+        callback: @escaping (UploadResult) -> Void
+    )
+}
+
+struct PingUploadRequest {
+    let documentId: String
+    let url: String
+    let data: [UInt8]
+    let headers: HeadersList
+    let uploaderCapabilities: [String]
+
+    init(request: PingRequest, endpoint: String) {
+        self.documentId = request.documentId
+        self.url = endpoint + request.path
+        self.data = request.body
+        self.headers = request.headers
+        self.uploaderCapabilities = request.uploaderCapabilities
+    }
+}
+
+public struct CapablePingUploadRequest {
+    let request: PingUploadRequest
+
+    init(_ request: PingUploadRequest) {
+        self.request = request
+    }
+
+    func capable(f: (_ uploaderCapabilities: [String]) -> Bool)
+        -> PingUploadRequest? {
+        if f(request.uploaderCapabilities) {
```

Comment:
Could we find another name for that variable? `f` is a bit short and I'm not sure what it means here

Reply:
Yes, I fear I may have just cargo-culted the `f` parameter naming and it can definitely be more descriptive.